### ISSUE TITLE
Docker daemon restarting in our large-cluster performance tests

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#62456]
We started seeing failures in our 2k-node and 5k-node performance tests recently:
- https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-large-performance (e.g run 78)
- https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-performance

@wojtek-t noticed that apiserver went down in between the test for ~1 minute. We then looked into the logs of other components and all of them seemed to have died during that period. And from the [docker logs](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/78/artifacts/gce-scale-cluster-master/docker.log) the daemon was restarted at that time:

```
Stopping Docker Application Container Engine...
time="2018-04-05T15:20:14.742673061Z" level=info msg="Processing signal 'terminated'"
time="2018-04-05T15:20:14.75474265Z" level=info msg="stopping containerd after receiving terminated"
docker daemon exited
```

Also, kubelet was restarted around the same time. All of this seems to suggest that this issue is something low level (i.e related to OS or sth) and I suspect the reason is COS image update that I remember was done just around that time (i.e 5th march). Specifically, this PR https://github.com/kubernetes/kubernetes/pull/61072 which changed it.

Master logs for that run are [here](http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/78/artifacts/gce-scale-cluster-master) and this bad behavior happened b/w 15:19 and 15:20.

/cc @kubernetes/sig-scalability-bugs @kubernetes/sig-node-bugs @dchen1107 @yujuhong

**Comment by shyamjvs** [comment]
I'm now trying to verify the hypothesis that it's because of the COS change by changing the version to the older one for our tests.

**Comment by shyamjvs** [comment]
Actually, I'm seeing that PR https://github.com/kubernetes/kubernetes/pull/61072 updated the cos versions to `cos-stable-65` for config-default and for kubemark, but to `cos-beta-66` for config-test (which is what we're performing our OSS-testing against). This IIUC means we've bypassed testing for cos-stable-65.

@filbranden Am I missing sth or was this intentional? Because I'd expect that we test the version used for prod.

**Comment by filbranden** [comment]
Hi @shyamjvs,

Are you using the setting that is coming from `cluster/gce/config-test.sh`? If you want prod, shouldn't you use `cluster/gce/config-default.sh` instead? (Or perhaps `cluster/kubemark/gce/config-default.sh`, since I imagine that is used in other benchmarks too...)

**Comment by shyamjvs** [comment]
> Are you using the setting that is coming from cluster/gce/config-test.sh?

Yes - just like ~all our OSS tests (not just scalability ones) are.

> If you want prod, shouldn't you use cluster/gce/config-default.sh instead?

Well, yes. But the point I mention is that we should be testing stuff that we use for prod - as we don't want to be testing one thing and running sth else in prod :)
In this particular case (where we jumped from cos-63 to cos-beta-66), we changed the prod config to cos-65 without ever having it covered by our tests.

**Comment by filbranden** [comment]
The reason for going to M66 in `config-test.sh` is that we also want to test what we *want* to run in prod. We're launching GKE 1.10 EAP with M66 and intend to stay on that milestone of COS for the life cycle of 1.10.

If you want to run both *current* prod and *next* prod, then you should set up two separate tests, one with `config-default.sh` and one with `config-test.sh`. That would really be ideal.

(Oh yes and I know in the past we've been slow to change these files... We're planning to do that more often now, both to reflect our current production and to get some early testing before we adopt specific versions in our production. For instance, before we launch GKE 1.11, we'll probably update `config-test.sh` to reflect the COS version we're planning to launch it on...)

Cheers,
Filipe

**Comment by shyamjvs** [comment]
I understand the requirement to test current prod and next prod too. All I'm saying is that we shouldn't test next prod without testing current prod - i.e in this particular case, we seem to have *never* used cos-stable-65 in testing because the switch was directly made from stable-63 -> beta-66 in config-test.sh. That said, let me know if I'm missing sth here.

**Comment by filbranden** [comment]
As I said, in the past we were slow to update this... But now we decided we want to test both, so we decided `config-test.sh` is the best place to test *next*.

We'll keep the pace of updating these in the future. So you can expect the next change is for `config-test.sh` to test whatever COS version we expect to go into GKE 1.11, while `config-default.sh` will then move to M66. So long term your expectations that every version will go through each of `config-test.sh` and `config-default.sh` will hold. We just had a *blip* here because we didn't play too close attention to these files, so now we had to skip ahead a bit to catch up...

Cheers,
Filipe

**Comment by filbranden** [comment]
Regarding the original issue... Is the problem you're seeing on `cos-stable-65` or `cos-beta-66`?

I'd expect `cos-stable-65` not to cause too much trouble here... Since it ships the same kernel family (4.4) and same Docker version as `cos-stable-63`...

Regarding `cos-beta-66`, the Docker version is still the same but the kernel is newer (based on 4.14). If you're seeing regressions there, I'd be curious to know that, since we're looking at that kernel and expecting it will fix issues we've seen on 4.4. Can you please clarify?

**Comment by shyamjvs** [comment]
> so now we had to skip ahead a bit to catch up

Another option could've been to make 2 quick updates (within 1 release). Where the 1st update would be from 63 -> 65 in test and the 2nd would be the one bringing it to the current state. That said, the current is probably fine.

> Is the problem you're seeing on cos-stable-65 or cos-beta-66?

cos-beta-66

> Can you please clarify?

Even though a strong suspicion for this regression is on the cos image update, I think we should still confirm that's the case. So I made https://github.com/kubernetes/test-infra/pull/7668 to check that reverting back to 63 helps. Once we're sure, we can then try to understand (together with sig-node) what exactly is causing it.

**Comment by filbranden** [comment]
Awesome, thanks @shyamjvs !

**Comment by yujuhong** [comment]
> As I said, in the past we were slow to update this... But now we decided we want to test both, so we decided config-test.sh is the best place to test next.

We actually used to rigorously update the images used in theses scripts, but in the end, the versions in OSS scripts fell behind what are actually in use in GKE due to various reasons (e.g., security patches and COS EOL problems). Given that cos m65 has already been in use in GKE for a while, it should be be safe to update these scripts to match it.

> Even though a strong suspicion for this regression is on the cos image update, I think we should still confirm that's the case. So I made kubernetes/test-infra#7668 to check that reverting back to 63 helps. Once we're sure, we can then try to understand (together with sig-node) what exactly is causing it.

@shyamjvs would it be possible to keep the VM from the failed test to inspect?

**Comment by shyamjvs** [comment]
> would it be possible to keep the VM from the failed test to inspect?

@yujuhong Currently I've changed our tests to run with older one (i.e cos-63) in https://github.com/kubernetes/test-infra/pull/7668. So I expect such failure to not happen. After that's confirmed, we can manually experiment with beta-66 again and it should be possible to keep it alive.

**Comment by shyamjvs** [comment]
So the latest runs of our 5k and 2k-node performance tests are green and I can't find any master restarts. Ref:
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/139
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/81
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/82

So my suspicion about issue being with cos is now verified :)

@yujuhong From the internal bug, it seems like this issue is reproducible. Let me know if you still want me to up a large cluster and keep it alive.

**Comment by shyamjvs** [comment]
To clarify - the way I confirmed the above was by checking from docker logs that it wasn't killed/restarted and from the serial port logs that no kernel panic had occurred.

**Comment by yujuhong** [comment]
> @yujuhong From the internal bug, it seems like this issue is reproducible. Let me know if you still want me to up a large cluster and keep it alive.

@shyamjvs, looks like the COS team were able to reproduce this issue. Will let you know otherwise. Thanks!

**Comment by shyamjvs** [comment]
@yujuhong @filbranden Should we bump down the version (if so, to which one) in our config-default/test files until a patch fix is available?
